vcd/resource_vcd_catalog_media.go

```diff
@@ -132,39 +171,74 @@ func resourceVcdMediaCreate(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("error waiting for task to complete: %+v", err)
 	}
 
-	d.SetId(catalogName + ":" + mediaName)
-
 	log.Printf("[TRACE] Catalog media created: %#v", mediaName)
 
 	err = createOrUpdateMediaItemMetadata(d, meta)
 	if err != nil {
 		return fmt.Errorf("error adding media item metadata: %s", err)
 	}
 
+	//sets Id in findCatalogItem func - there isn't ID with media type
 	return resourceVcdMediaRead(d, meta)
 }
 
 func resourceVcdMediaRead(d *schema.ResourceData, meta interface{}) error {
 	vcdClient := meta.(*VCDClient)
 
-	_, vdc, err := vcdClient.GetOrgAndVdc("", "")
+	org, err := vcdClient.GetAdminOrgFromResource(d)
 	if err != nil {
 		return fmt.Errorf(errorRetrievingOrg, err)
 	}
 
-	mediaItem, err := vdc.FindMediaImage(d.Get("name").(string))
-	if err != nil || mediaItem == (govcd.MediaItem{}) {
-		log.Printf("[DEBUG] Unable to find media item: %s", err)
+	catalog, err := org.GetCatalogByName(d.Get("catalog").(string), false)
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find catalog.")
+		return fmt.Errorf("unable to find catalog: %s", err)
+	}
+
+	identifier := d.Id()
+
+	if identifier == "" {
+		identifier = d.Get("name").(string)
+	}
+
+	media, err := catalog.GetMediaByNameOrId(identifier, false)
+	if govcd.IsNotFound(err) {
+		log.Printf("unable to find media with ID %s: %s. Removing from state", identifier, err)
+		d.SetId("")
+		return nil
+	}
+	if err != nil {
+		log.Printf("[DEBUG] Unable to find media: %s", err)
+		return err
+	}
+
+	d.SetId(media.Media.ID)
+
+	mediaRecord, err := catalog.QueryMedia(d.Get("name").(string))
+	if err != nil {
+		log.Printf("[DEBUG] Unable to query media: %s", err)
```

Comment:
This double request makes the code difficult to read.
I think we should change `GetMediaBy(Name|Id|Href)` to return a full media record, meaning that inside these functions we call `QueryMedia` and return a single structure containing everything we need.

Reply:
I would agree on having wrapper, cause not always you need `all the data` and thus create two request when you need one. And yet more votes needed for that, IMHO I am ok with this approach.